### PR TITLE
feat(query): fix distinct aggregate function

### LINKED ISSUE #7743: count distinct bug
```SELECT COUNT(DISTINCT SearchPhrase) FROM hits ```  return 3744 rows
``` SELECT COUNT(DISTINCT SearchPhrase) FROM hits``` return 3756， （diff num for each query）

how to reproduce:
-----

1. pull this pr  and start standalone
2.  serve http file local

```
curl https://repo.databend.rs/dataset/stateful/hits_100k.tsv
python -m  http.server 80
```
3. sql

```
COPY INTO hits FROM 'http://loaclhoust/hits_100k.tsv' FILE_FORMAT = ( type = 'tsv' record_delimiter = '\n' skip_header = 1 )
```

-----

``` SELECT COUNT(DISTINCT SearchPhrase) FROM  (select SearchPhrase from hits order by SearchPhrase) ```   is 3744

`set max_threads=1`  before copy fix it, so it is a bug COUNT DISTINCT with multi partition

_Originally posted by @youngsofun in https://github.com/datafuselabs/databend/issues/7613#issuecomment-1251790261_

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Fixes #7743

### CODE REVIEW

**Review by zhang2014 — APPROVED** [review]

File: src/query/functions/src/aggregates/aggregate_distinct_state.rs [hunk 1]
```diff
@@ -171,31 +179,58 @@ impl DistinctStateFunc<DataGroupValues> for AggregateDistinctState {
...
 
+impl AggregateDistinctStringState {
+    #[inline]
+    fn insert_and_materialize(&mut self, key: &KeysRef) {
+        let entity = self.set.insert_key(key, &mut self.inserted);
+        if self.inserted {
+            let data = unsafe { key.as_slice() };
+
+            let holder = self.holders.last_mut().unwrap();
+            if holder.may_resize(data.len()) {
```
Comment (zhang2014) [inline src/query/functions/src/aggregates/aggregate_distinct_state.rs:190]: There will be memory fragmentation here, please add comment for refactor in future

Comment (sundy-li) [reply src/query/functions/src/aggregates/aggregate_distinct_state.rs:190]: ok

File: src/query/functions-v2/src/aggregates/aggregate_distinct_state.rs [hunk 2]
```diff
@@ -148,26 +153,61 @@ impl DistinctStateFunc<DataGroupValue> for AggregateDistinctState {
...
     fn deserialize(&mut self, reader: &mut &[u8]) -> Result<()> {
-        self.holder = deserialize_from_slice(reader)?;
-        self.set = HashSet::with_capacity(self.holder.len());
-
-        for index in 0..self.holder.len() {
-            let data = unsafe { self.holder.index_unchecked(index) };
-            let key = KeysRef::create(data.as_ptr() as usize, data.len());
-            self.set.insert(key);
+        self.holders = deserialize_from_slice(reader)?;
+        self.set = CommonHashSet::with_capacity(self.holders.iter().map(|h| h.len()).sum());
```
Comment (RinChanNOWWW) [inline src/query/functions-v2/src/aggregates/aggregate_distinct_state.rs:203]: `self.holders.iter().map(|h| h.len()).sum()`

Can use `fold`?

**Review by RinChanNOWWW — APPROVED** [review]

File: src/query/functions-v2/src/aggregates/aggregate_distinct_state.rs (hunk 2 above)
Comment (sundy-li) [reply src/query/functions-v2/src/aggregates/aggregate_distinct_state.rs:203]: Yes, but they are pretty similar way...

**Pull request merged**